prow/kube/config.go

```diff
@@ -34,64 +34,110 @@ import (
 // contexts are found.
 func LoadClusterConfigs(kubeconfig, buildCluster string) (configurations map[string]rest.Config, defaultContext string, err error) {
```

Comment:
Given all the trouble we've been having here, can we have a less aggressive refactor?

This file is absolutely testable as is. The in cluster configuration should be largely irrelevant as it should only add contexts, and we do not really care if there are extra contexts, but we need to validate that the ones we expect a) exist and b) have the right values.

We should only do this kind of refactor after we've added the right TestLoadClusterConfigs tests to be confident that we have these values correct.

Reply:
I don't see how to write a test for this function before any changes ... InClusterConfig will change the output state based on where you're calling the test/

Reply:
Whether or not InClusterConfig is not irrelevant, it's critical to anything working in the service cluster...

Reply:
We should operate in one of two modes:
* --build-cluster is defined and we only use these values
* --build-cluster is not defined and we use the modern behavior.

Right now this is a fairly serious change in behavior from before:
* We always load the in cluster config
* We always load the kubeConfigLoader (when kubeconfig=="" we still call `clientcmd.NewDefaultClientConfigLoadingRules()` which will pick up implicit configs shoved in a variety of places, which distros can and do inject (I believe GKE does this)

This does not seem like the right first step.

Reply:
Ok, if we want that disjoint set of options, we should do a different set of work to get there. Agreed. I'll do that.

Reply:
Wait no, you don't have the correct state of the world. Currently -- without this patch -- we already always use in-cluster config for the service cluster _and_ as the default for the build cluster.

Reply:
See L115 on the left side. Anything defined in the build-cluster file will override other values.

Reply:
I will write unit tests that enforce that behavior before this PR and show that this PR does not change that behavior.